src/components/scenes/Fio/FioCreateHandleScene.tsx

```diff
@@ -37,128 +34,119 @@ const asRegisterFailedRes = asObject({
   error: asString
 })
 
+const asFreeFioDomain = asObject({
+  domain: asString,
+  free: asValue(true)
+})
+
 export const FioCreateHandleScene = ({ navigation }: Props) => {
   const theme = useTheme()
   const styles = getStyles(theme)
   const dispatch = useDispatch()
   const accountUserName = useSelector(state => state.core.account.username)
-  const domainStr = `@${config.appName.toLowerCase()}`
+  const fioPlugin = useSelector(state => state.core.account.currencyConfig.fio)
+
+  const { freeRegApiToken, freeRegRefCode } = typeof ENV.FIO_INIT === 'object' ? ENV.FIO_INIT : { freeRegApiToken: null, freeRegRefCode: null }
 
   const [wallet, setWallet] = React.useState<EdgeCurrencyWallet | undefined>()
+  const [domainStr, setDomainStr] = React.useState<string>('')
   const [fioHandle, setFioHandle] = React.useState<string>('')
   const [errorText, setErrorText] = React.useState<string>()
+
   const inputRef = React.useRef<TextInput>(null)
+  const mounted = React.useRef<boolean>(true)
 
   const handleChangeFioHandle = useHandler((userInput: string) => {
     // Clean the userInput:
     userInput = userInput.replace(domainStr, '').trim()
 
-    // ASCII a-z 0-9. Remove all non-alphanumeric characters, convert to
-    // lowercase
-    userInput = userInput.replace(/[^a-z0-9]/gi, '').toLowerCase()
-
     // Dash '-' allowed, but cannot be first or last character
-    if (userInput.length > 1) {
-      userInput = userInput.charAt(0) + userInput.slice(1, -1).replace('-', '') + userInput.slice(-1)
-    } else if (userInput === '-') {
-      userInput = ''
-    }
+    userInput = userInput.charAt(0).replace('-', '') + userInput.slice(1, -1) + userInput.slice(-1).replace('-', '')
 
-    setFioHandle(`${userInput}`)
+    // ASCII a-z 0-9. Remove all non-alphanumeric characters, convert to
+    // lowercase. Allow dashes as they were cleaned above
+    userInput = userInput.replace(/[^a-z0-9-]/gi, '').toLowerCase()
+
+    setFioHandle(userInput)
   })
 
   const handleRegisterPress = async () => {
-    // Register button is disabled if wallet creation isn't finished yet
+    // Check if ENV is configured with required keys and codes
+    if (freeRegApiToken == null || freeRegRefCode == null) {
+      if (!mounted.current) return
+      setErrorText(s.strings.fio_register_handle_error)
+      return
+    }
+
+    // Register button is disabled if wallet ctreation isn't finished yet.
+    // Shouldn't happen.
     if (wallet == null) return
+
+    // Check if the handle is already registered
     const fioName = `${fioHandle}${domainStr}`
+    if (!(await fioPlugin.otherMethods.isFioAddressValid(fioName))) {
```

Comment:
This does NOT check if address is available. Only that it's a valid format. My comment explicitly said to use `otherMethods.validateAccount(accountName)`